lib/src/user_config.rs

```diff
@@ -0,0 +1,245 @@
+    HomeDirNotFound,
+    /// Failed to read / write to the specified path
+    #[error(transparent)]
+    PathError(#[from] PathError),
+    /// Failed to load the private key.
+    #[error(transparent)]
+    PrivateKeyError(#[from] ed25519_dalek::pkcs8::Error),
+    /// Failed to load the public key.
+    #[error(transparent)]
+    PublicKeyError(#[from] ed25519_dalek::pkcs8::spki::Error),
+
+    /// Failed to decoode the user configuration file.
+    #[error(transparent)]
+    DecodeError(#[from] prost::DecodeError),
+
+    /// Missing the user configuration file.
+    #[error(
+        "This repo was generated by an old version of jj and does not have secure user \
+         configuration"
+    )]
+    LegacyRepo,
+}
+
+/// A config file represented by a string instead of backed by a file.
+pub trait ConfigType: prost::Message + Default + Clone + Debug {
+    /// The name of this type.
+    fn kind() -> &'static str;
+    /// The filename of the configuration file to write.
+    fn filename() -> &'static str;
+    /// The string containing the config.
+    fn config(&self) -> &str;
+}
+
+impl ConfigType for RepoConfig {
+    fn kind() -> &'static str {
+        "repo"
+    }
+
+    fn filename() -> &'static str {
+        "config.binpb"
+    }
+
+    fn config(&self) -> &str {
+        &self.config
+    }
+}
+
+impl ConfigType for WorkspaceConfig {
+    fn kind() -> &'static str {
+        "workspace"
+    }
+
+    fn filename() -> &'static str {
+        "workspace-config.binpb"
+    }
+
+    fn config(&self) -> &str {
+        &self.config
+    }
+}
+
+/// The result of a read of user configuration.
+#[derive(Clone, Debug)]
+pub enum UserConfig<T: ConfigType> {
+    /// We had the correct signature.
+    Trusted(T),
+    /// Repository has moved from one location to another.
+    RepoMoved {
+        /// The old path to the repo, encoded as a string.
+        from: String,
+        /// The new path to the repo, encoded as a string.
+        to: String,
+        /// The configuration that the repo tried to use.
+        config: T,
+    },
+
+    /// Repository had the wrong signature.
+    InvalidSignature(T),
+}
+
+impl<T: ConfigType> UserConfig<T> {
+    /// Retrieves the underlying config, regardless of the level of trust.
+    pub fn get(&self) -> &T {
+        match self {
+            Self::Trusted(config)
+            | Self::RepoMoved { config, .. }
+            | Self::InvalidSignature(config) => config,
+        }
+    }
+}
+
+/// Returns a signing key associated with the current user.
+/// Generates a key if we don't yet have one.
+fn signing_key() -> Result<SigningKey, UserConfigError> {
+    let strategy = match etcetera::choose_base_strategy() {
+        Ok(s) => s,
+        Err(_) => return Err(UserConfigError::HomeDirNotFound),
+    };
+    let key_dir = strategy.data_dir().join("jj/config/keys");
+    let signing_key = key_dir.join("ed25519.der");
+    let verifying_key = key_dir.join("ed25519.der.pub");
```

Comment:
I don't think we need public-key cryptography to sign and verify user's private data by his key. What we need is something like `hash(private_key + path + config_data)` but is considered secure. I _think_ some MAC algorithm can be used.
https://en.wikipedia.org/wiki/Message_authentication_code

I'm not an expert, but pubkey-based algorithm would be computationally more expensive than symmetric-key algorithms or cryptographic hashing.

Reply:
I double-checked with a security expert, they say that you're right and that asymmetric keys aren't improving anything.

They also mentioned that specifically, "HMAC_SHA256 would be a sensible choice"